programs/clearing_house/src/controller/orders.rs

```diff
@@ -570,6 +486,241 @@ pub fn fill_order(
+) -> ClearingHouseResult<(u128, bool)> {
+    let market_index = user.orders[user_order_index].market_index;
+
+    let user_checkpoint = checkpoint_user(user, market_index, Some(user_order_index))?;
+    let maker_checkpoint = if let Some(maker) = maker {
+        let maker_order_index = maker_order_index.ok_or(ErrorCode::MakerOrderNotFound)?;
+        Some(checkpoint_user(
+            maker,
+            market_index,
+            Some(maker_order_index),
+        )?)
+    } else {
+        None
+    };
+    let filler_checkpoint = if let Some(filler) = filler {
+        Some(checkpoint_user(filler, market_index, None)?)
+    } else {
+        None
+    };
+
+    let market_checkpoint = clone(market_map.get_ref(&market_index)?.deref());
+
+    let fulfillment_methods =
+        determine_fulfillment_methods(&user.orders[user_order_index], false, now)?;
+
+    if fulfillment_methods.is_empty() {
+        return Ok((0, false));
+    }
+
+    let mut base_asset_amount = 0_u128;
+    let mut potentially_risk_increasing = false;
+    let mut order_records: Vec<OrderRecord> = vec![];
+    for fulfillment_method in fulfillment_methods.iter() {
+        if user.orders[user_order_index].status != OrderStatus::Open {
+            break;
+        }
+
+        let mut market = market_map.get_ref_mut(&market_index)?;
+
+        let (_base_asset_amount, _potentially_risk_increasing) = match fulfillment_method {
+            FulfillmentMethod::AMM => fulfill_order_with_amm(
+                state,
+                user,
+                user_order_index,
+                market.deref_mut(),
+                oracle_map,
+                mark_price_before,
+                now,
+                valid_oracle_price,
+                user_key,
+                filler_key,
+                filler,
+                &mut order_records,
+            )?,
+            FulfillmentMethod::Match => fulfill_order_with_match(
+                market.deref_mut(),
+                user,
+                user_order_index,
+                user_key,
+                maker.as_deref_mut().unwrap(),
+                maker_order_index.unwrap(),
+                maker_key.unwrap(),
+                filler.as_deref_mut(),
+                filler_key,
+                now,
+                &state.fee_structure,
+                oracle_map,
+                &mut order_records,
+            )?,
+        };
+
+        potentially_risk_increasing = potentially_risk_increasing || _potentially_risk_increasing;
+        base_asset_amount = base_asset_amount
+            .checked_add(_base_asset_amount)
+            .ok_or_else(math_error!())?;
+    }
+
+    let meets_initial_margin_requirement =
+        meets_initial_margin_requirement(user, market_map, bank_map, oracle_map)?;
+
+    if meets_initial_margin_requirement {
+        for order_record in order_records {
+            emit!(order_record)
+        }
+    } else {
+        revert_to_checkpoint(user, user_checkpoint)?;
```

Comment:
what's the reason to go through all this trouble saving/reverting user and market checkpoints instead of just throwing an error? wouldn't an error automatically revert all the state changes in the transaction?

Reply:
i guess this way you are able to cancel the order if it filling leaves the user in a dangerous account position?

Reply:
yeah thats one reason. the second reason is that it makes the keeper logic easier of chain because it doesnt need to check user margin requirment